compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1166,26 +1168,38 @@ private BAttachedFunction getNextFunc(BObjectType iteratorType) {
 
     public BType inferRecordFieldType(BRecordType recordType) {
         List<BField> fields = recordType.fields;
+        BUnionType unionType = BUnionType.create(null);
 
-        // If there are no fields in the record, return the rest field type as the inferred type.
-        if (fields.isEmpty()) {
-            return recordType.restFieldType;
+        if (recordType.restFieldType.tag != TypeTags.NONE) {
```

Comment:
Shall we check
```suggestion
        if (!recordType.sealed) {
```
instead?